src/Text/Pandoc/Writers/JATS.hs

```diff
@@ -54,6 +54,64 @@ import Text.TeXMath
 import qualified Text.Pandoc.Writers.AnnotatedTable as Ann
 import qualified Text.XML.Light as Xml
 
+-- Create a nested dictionary data structure that contains
+-- the Contributor Role Taxonomy (CRediT). This is useful
+-- for generating JATS that annotate contributor roles
+creditNames :: M.Map Text Text
+creditNames = M.fromList [
+    ("conceptualization", "Conceptualization"),
+    ("data-curation", "Data curation"),
+    ("formal-analysis", "Formal analysis"),
+    ("funding-acquisition", "Funding acquisition"),
+    ("investigation", "Investigation"),
+    ("methodology", "Methodology"),
+    ("project-administration", "Project administration"),
+    ("resources", "Resources"),
+    ("software", "Software"),
+    ("supervision", "Supervision"),
+    ("validation", "Validation"),
+    ("visualization", "Visualization"),
+    ("writing-original-draft", "Writing – original draft"),
+    ("writing-review-editing", "Writing – review & editing")]
+
+addCreditNames :: Context a -> Context a
```

Comment:
This type might need to be changed to `Context Text -> Context Text`.

Reply:
Okay I think this helped. Now the error is inside the other functions:

```haskell
src/Text/Pandoc/Writers/JATS.hs:79:8: error: [GHC-39999]
    • No instance for ‘Text.DocTemplates.Internal.FromContext
                         Text (M.Map Text Text)’
        arising from a use of ‘getField’
    • In the expression: getField "roles" context
      In the expression:
        case getField "roles" context of
          Nothing -> context
          Just roles -> resetField "roles" (map addCreditName roles) context
      In an equation for ‘addCreditNames’:
          addCreditNames context
            = case getField "roles" context of
                Nothing -> context
                Just roles -> resetField "roles" (map addCreditName roles) context
   |
79 |   case getField "roles" context of
   |        ^^^^^^^^

```

Reply:
Look at https://hackage.haskell.org/package/doctemplates-0.11.0.1/docs/Text-DocTemplates.html

You can see that there is a ToContext instance for ToContext b => Map Text b,
but there isn't a similar FromContext instance!  Argh.

Probably I should just add one.  Stand by.

Reply:
Well actually we may be able to work around this by operating directly with a Val type instead of a Map.
```hs
addCreditName :: Val Text -> Val Text
addCreditName val =
  case val of
    MapVal ctx ->
      case getField "credit-name" ctx of
        Just _ -> val
        Nothing -> maybe val (MapVal . resetField "credit-name")
          (getFiled "credit-id" ctx >>= flip M.lookup creditNames)
    _ -> val
```
(UNTESTED! But I think something along these lines will work, without the need for changes in doctemplates.)

Reply:
I tried a couple variations on this. The code in https://github.com/jgm/pandoc/pull/10153#discussion_r1776077467 gave this error:

```haskell
src/Text/Pandoc/Writers/JATS.hs:95:40: error: [GHC-83865]
    • Couldn't match type: Context a0 -> Context a0
                     with: Context Text
      Expected: Text -> Context Text
        Actual: Text -> Context a0 -> Context a0
    • Probable cause: ‘resetField’ is applied to too few arguments
      In the second argument of ‘(.)’, namely ‘resetField "credit-name"’
      In the second argument of ‘maybe’, namely
        ‘(MapVal . resetField "credit-name")’
      In the expression:
        maybe
          val (MapVal . resetField "credit-name")
          (getField "credit-id" ctx >>= flip M.lookup creditNames)
   |
95 |         Nothing -> maybe val (MapVal . resetField "credit-name")
   |     
```

Unfortunately, I don't think it's very effective working this way and I'm not sure how to push this forward. If we agree on the base ideas (extending the documented data model, how the templating for JATS changes, and how the tests look) then I would say we should split this PR into two parts - one that introduces those changes, then a second as a follow-up that uses some haskell magic to help when data is incomplete (i.e. the changes we're talking about in this thread)

@jgm do you think that would work? Otherwise I fear that I might not be able to get the job done and am sucking up too much of your open source time, and/or this thread becomes impossible to follow

Reply:
As the message indicates, `resetField` doesn't have enough parameters.  Just a dumb mistake on my part.
```hs
        Nothing -> maybe val (MapVal . resetField "credit-name")
          (getField "credit-id" ctx >>= flip M.lookup creditNames)
```
should be
```hs
        Nothing -> maybe val (MapVal . (\x -> resetField "credit-name" x val)
          (getField "credit-id" ctx >>= flip M.lookup creditNames)
```
Or, more verbose but easier to follow:
```hs
        Nothing ->
           case getField "credit-id" ctx of
             Nothing -> val
             Just creditId ->
               case M.lookup creditId creditNames of
                 Nothing -> val
                 Just creditName -> resetField "credit-name" creditName val
```

Reply:
alright, I think we're making progress, but now it seems that the way getField is defined is causing the issue:

```haskell
src/Text/Pandoc/Writers/JATS.hs:93:12: error: [GHC-39999]
    • Ambiguous type variable ‘b0’ arising from a use of ‘getField’
      prevents the constraint ‘(Text.DocTemplates.Internal.FromContext
                                  Text b0)’ from being solved.
      Probable fix: use a type annotation to specify what ‘b0’ should be.
      Potentially matching instances:
        instance Text.DocTemplates.Internal.TemplateTarget a =>
                 Text.DocTemplates.Internal.FromContext a a
          -- Defined in ‘Text.DocTemplates.Internal’
        instance Text.DocTemplates.Internal.TemplateTarget a =>
                 Text.DocTemplates.Internal.FromContext a (Doc a)
          -- Defined in ‘Text.DocTemplates.Internal’
        ...plus two others
        (use -fprint-potential-instances to see them all)
    • In the expression: getField "credit-name" ctx
      In the expression:
        case getField "credit-name" ctx of
          Just _ -> val
          Nothing
            -> case getField "credit-id" ctx of
                 Nothing -> val
                 Just creditId
                   -> case M.lookup creditId creditNames of
                        Nothing -> ...
                        Just creditName -> ...
      In a case alternative:
          MapVal ctx
            -> case getField "credit-name" ctx of
                 Just _ -> val
                 Nothing
                   -> case getField "credit-id" ctx of
                        Nothing -> val
                        Just creditId -> ...
   |
93 |       case getField "credit-name" ctx of
   |            ^^
```

Reply:
>       Probable fix: use a type annotation to specify what ‘b0’ should be.

This is your hint.  `getField` is polymorphic, so we may need to explicitly specify the type of the intended result, if it's not pinned down by other uses.

in this case, `getField "credit-name" ctx` is ambiguous because we don't do anythnig with the `Just _` return value ,so it could be any type.  Thus, we need to say what the type is that we expect:

```
Just (_ :: Text) -> val
```
if we want to leave it alone if it's a text value, or
```
Just (_ :: Val Text) -> val
```
if we want to leave it alone if it's any kind of value.

Reply:
Woohoo! It has a few warnings but managed to compile! I added a corresponding test to make sure that it works (i.e., if there's a `credit-id` and no `credit-name`, add the right `credit-name`.) This isn't passing yet, so this needs some more invesigation

Reply:
At line 92 you match Just (MapVal _) but you don't match anything else.
You need a case for the catch-all "other" clause: `_ -> val`.
See my original example at https://github.com/jgm/pandoc/pull/10153#discussion_r1776077467

Reply:
hi @jgm sorry I had to step away from this because I felt like I had hit a wall, but I realized an issue with the way my filtering was working was it didn't match the data structure (I was looking for a roles list, but the only thing that was available was an authors list, and there was a roles list in each)

I have (I hope) a final issue with the compiler, again related to me not understanding the typing system enough to annotate the cases correctly. Could you please give me another hint?

```
src/Text/Pandoc/Writers/JATS.hs:79:8: error: [GHC-39999]
    • No instance for ‘Text.DocTemplates.Internal.FromContext
                         Text (Context Text)’
        arising from a use of ‘getField’
    • In the expression: getField "authors" context
      In the expression:
        case getField "authors" context of
          Nothing -> error "No Authors!"
          Just authors
            -> resetField
                 "authors" (map addCreditNamesToAuthor authors) context
      In an equation for ‘addCreditNames’:
          addCreditNames context
            = case getField "authors" context of
                Nothing -> error "No Authors!"
                Just authors
                  -> resetField
                       "authors" (map addCreditNamesToAuthor authors) context
   |
79 |   case getField "authors" context of
   |        ^^
```